[Asset] Prevent flooding pimcore_asset_update message queue transport with faulty assets triggered by previews in admin ui

brusch opened the pull request:
+ some further optimizations 😉 

How to reproduce: 
If you have a faulty asset from type video or document and you request a preview in the admin ui, it's missing the needed meta-data stored as custom settings (width, height, duration, pageCount), so the asset is getting pushed into the pimcore_asset_update queue for processing. Since the asset is faulty, the needed meta-data can't be extracted, so it won't be available anytime, so the asset get's pushed into the queue again and again.

brusch committed:
[Asset] Prevent flooding pimcore_asset_update message queue transport with faulty assets triggered by previews in admin ui

brusch committed:
Apply php-cs-fixer changes

brusch committed:
[Asset] Implemented locking for pimcore_asset_update message queue transport

brusch committed:
Apply php-cs-fixer changes

brusch commented:
Also added locking, as there's no need to have duplicates in the queue, as the message contains only the asset ID. 
Locking seems to me the most efficient and simple way of avoiding duplicated in the queue.

brusch committed:
Merge remote-tracking branch 'origin/11.5' into asset-update-processing-fix

# Conflicts:
#	lib/Messenger/Handler/AssetUpdateTasksHandler.php

brusch committed:
making SonarQube happy

brusch committed:
Apply php-cs-fixer changes

brusch commented:
Created a related PR for admin ui: https://github.com/pimcore/admin-ui-classic-bundle/pull/998

herbertroth commented:
Could you please assign the required reviewer? This PR was moved to "In Review" status but has no reviewers assigned.

herbertroth commented:
@kingjia90 Could you please assign the required reviewer?

kingjia90 commented:
> @kingjia90 Could you please assign the required reviewer?

haha 😄 i'd suggest to replace with a Bot account, this automated message feels weird sometimes

herbertroth commented:
😅 Haha, looks like my “automation” needs a bit more training… guess I’ll have to put it back on the workbench again! 🛠️🤖

kingjia90 commented on models/Asset/Video.php:
```diff
@@ -216,9 +220,16 @@ public function getDuration(): float|int|null
-        if (!$width || !$height) {
+
+        if ($width && $height) {
+            $dimensions = [
```
what about "reserving" the assignment of `$dimensions` to the output `$this->getDimensionsFromBackend();` 
```suggestion
            return [
```

remove `$dimensions = null;` on line 223 and have a  `return null` at the bottom (the implicit else all)

brusch replied:
But then we'd need another return `$dimensions` in the `getDimensionsFromBackend() / elseif` so nothing saved 😉

kingjia90 commented on models/Asset/Document.php:
```diff
@@ -75,6 +75,8 @@ public function processPageCount(string $path = null): bool
             Logger::error((string) $e);
             $this->setCustomSetting('document_page_count', 'failed');
+
+            return false;
```
This change may prevent this from saving in the backend search index
https://github.com/pimcore/pimcore/blob/381d758b7f5545111fa863743dc25a8b92e2a4fc/bundles/SimpleBackendSearchBundle/src/Command/SearchBackendReindexCommand.php#L84-L86

brusch replied:
It would still save the asset in the index, but you're right, the asset itself with the changed `document_page_count` custom setting won't get saved. -> fixed that in 5d94ddd54ddc277d277f2ebdc308c77099a55f49

kingjia90 commented on lib/Messenger/Handler/AssetUpdateTasksHandler.php:
```diff
@@ -51,51 +56,62 @@ public function __invoke(AssetUpdateTasksMessage $message): void
             $asset->setCustomSetting('duration', $duration);
-        } else {
-            $asset->removeCustomSetting('duration');
+            if ($dimensions = $asset->getDimensionsFromBackend()) {
```
initially though that we may not need a nested IF there

```suggestion
            $dimensions = $asset->getDimensionsFromBackend();
            $duration = $asset->getDurationFromBackend();
            
            if ($dimensions && $duration) {
                $asset->setCustomSetting('duration', $duration);
```

kingjia90 replied:
Since it looks like they are less "independent" from each other, what about having a new method to `getMetaDataFromBackend()` (or so) that returns an associative array `['dimension'=>null, 'duration' => null]`

If i got correctly, by looking at
https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/models/Asset/Video.php#L161-L191

It calls [getDefaultAdapter](https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/lib/Video.php#L61-L80) (which has class exists check and initializes a new Adapter class
) at least 4 times (when processing is not considered to be failed) that might cause memory issues


more specifically in https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/lib/Video.php#L52-L56 and getInstance https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/lib/Video.php#L39

brusch replied:
> initially though that we may not need a nested IF there

I went for nested if, since it doesn't make sense to read dimensions if reading the duration failed already so save resources.

kingjia90 replied:
yes, agreeing with the nested here, but wdyt about this https://github.com/pimcore/pimcore/pull/18679#discussion_r2371387229 ? it's needing 4 instantiation of the adapter, not strictly for this PR but might be causing memory issues

brusch replied:
@kingjia90 yeah, you're right, looking at it, I'd see this as a follow up optimization. 
I think it would also make sense to introduce an interface for adapters and get rid of the [abstract ](https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/lib/Video/Adapter.php#L19)instead. 

I'll create a follow up bc-break ticket for that.

brusch replied:
done #18702

kingjia90 commented on models/Asset/Document.php:
```diff
@@ -75,6 +75,8 @@ public function processPageCount(string $path = null): bool
         } catch (Exception $e) {
             Logger::error((string) $e);
             $this->setCustomSetting('document_page_count', 'failed');
+
```
If we are returning false, we likely could get rid of all the logic revolving `customSetting('document_page_count')` set/get `failed` eg. https://github.com/pimcore/pimcore/pull/18679/files#diff-f9779d6bf7f7d2cfcbd48f036980ea510e017943bf93f2edbcc968d60689a020R82

brusch replied:
`processPageCount()` is only for reading the page count directly from the file, the custom setting `processPageCount`  is meant to "cache" the page count for further use, e.g. by `getPageCount()`. 
So `getPageCount()` should be the public getter that should be used by developers in project scope.

kingjia90 reviewed (changes requested):
The idea and logic behind LGTM, just some inputs that i have noticed while looking at it

brusch committed:
Search backend reindex: fixed PDF saving after page processing

kingjia90 commented on lib/Messenger/Handler/AssetUpdateTasksHandler.php:
```diff
@@ -51,51 +56,62 @@ public function __invoke(AssetUpdateTasksMessage $message): void
+                $this->logger->warning(sprintf('Failed processing page count for document asset %s.', $asset->getId()));
+
+            } else {
+                $asset->getImageThumbnail(Asset\Image\Thumbnail\Config::getPreviewConfig())->generate(false);
```
This line moved here might be problematic in the cases where we don't want to process the count or the adapter doesn't support counting, but it is still able to process it for preview generation.

In other words, should be considering counting disabled by system config treated the same as a `UPDATE_TASK_PROCESSING_FAILED` case?

Notes/See also
https://github.com/pimcore/pimcore/blob/5d94ddd54ddc277d277f2ebdc308c77099a55f49/models/Asset/Document.php#L53-L55
https://github.com/pimcore/pimcore/blob/3913551f4729b08023c7e8c61d3ebc7b1c0366b6/models/Asset/Document.php#L221

brusch replied:
That's a valid point - thanks! 
Should be fixed now.

kingjia90 commented on lib/Messenger/Handler/AssetUpdateTasksHandler.php:
```diff
@@ -51,51 +56,62 @@ public function __invoke(AssetUpdateTasksMessage $message): void
     private function processVideo(Asset\Video $asset): void
     {
+        $asset->removeCustomSetting(Asset::CUSTOM_SETTING_UPDATE_TASK_PROCESSING_FAILED);
+        $failed = true;
```
```suggestion
        $save = false;
```

What about something more similar to the $save (former $savedNeeded )in processDocument but for [line 131](https://github.com/pimcore/pimcore/pull/18679/files#diff-f9779d6bf7f7d2cfcbd48f036980ea510e017943bf93f2edbcc968d60689a020R131)?

Beside refactoring `$failed`, we would need to check if `SphericalMetaData` and `embeddedMetaDataExtracted` aren't changed by comparing the result with the stored getCustomSetting

brusch replied:
`$failed` : in `processDocument()` it's different, because there are multiple tasks, so having `false` as default and set `$save` to `true` makes more sense and works with less code, not? 

meta-data: true, it's not related to the bug fix, but still, it is/was a bug 😉 fixed that as well

kingjia90 replied:
ahh ok, i wanted to prevent calling a `saveAsset()` at all when the CustomSetting data happens somehow to be just the same, but it's also fine like this with the `$failed` approach.
it's indeed more "complex" than document.

brusch committed:
Search backend reindex: do not process page count for reindexing - as there's no real benefit

brusch committed:
- removed page count dependency on thumbnail generation and text extraction for asset documents - I don't see the point for that
- [removed page count processing](https://github.com/pimcore/pimcore/pull/18679/commits/9b96ab1e6675cf10fd2e6368145ddd56124ec02b) in simple backend search -> same here, I don't see any benefit for that, you can't search for that value anyway
- removed video dimension dependeny on thumbnail generation for videos - same here, I don't see any dependency on that
- renamed `CUSTOM_SETTING_UPDATE_TASK_PROCESSING_FAILED`  to `CUSTOM_SETTING_PROCESSING_FAILED` to be more generic so it can be used for thumbnails as well in future

brusch commented:
@kingjia90 I did some further optimizations: 
- removed page count dependency on thumbnail generation and text extraction for asset documents - I don't see the point for that 
- [removed page count processing](https://github.com/pimcore/pimcore/pull/18679/commits/9b96ab1e6675cf10fd2e6368145ddd56124ec02b) in simple backend search -> same here, I don't see any benefit for that, you can't search for that value anyway 
- removed video dimension dependeny on thumbnail generation for videos - same here, I don't see any dependency on that
- renamed `CUSTOM_SETTING_UPDATE_TASK_PROCESSING_FAILED`  to `CUSTOM_SETTING_PROCESSING_FAILED` to be more generic so it can be used for thumbnails as well in future

brusch committed:
Apply php-cs-fixer changes

brusch committed:
removed unused method  Pimcore\Bundle\SimpleBackendSearchBundle\Command\SearchBackendReindexCommand::saveAsset()

brusch committed:
Apply php-cs-fixer changes

kingjia90 commented on lib/Messenger/Handler/AssetUpdateTasksHandler.php:
```diff
@@ -51,67 +56,77 @@ public function __invoke(AssetUpdateTasksMessage $message): void
-        } catch (Exception $e) {
-            $this->logger->warning($e->getMessage());
-        }
+        $asset->handleEmbeddedMetaData();
```
I believe it should be also applied to `processImage()` too https://github.com/pimcore/pimcore/blob/82ac5da745eba22606f7ac56499783bbe18ef6b2/lib/Messenger/Handler/AssetUpdateTasksHandler.php#L158-L162

kingjia90 commented on models/Asset.php:
```diff
@@ -1783,14 +1787,29 @@ public function clearThumbnail(string $name): void
+     */
+    public function addToUpdateTaskQueue(): void
+    {
+        if (!$this->getCustomSetting(self::CUSTOM_SETTING_PROCESSING_FAILED)) {
```
By looking at this, should `$asset->removeCustomSetting(Asset::CUSTOM_SETTING_PROCESSING_FAILED);` be added it here near by too https://github.com/pimcore/pimcore/blob/295879e0fc3b47cf0c406d60e9bf20904bc0e767/models/Asset/Document.php#L37 ?

brusch replied:
I've moved that to the beginning of `\Pimcore\Messenger\Handler\AssetUpdateTasksHandler::__invoke()` I don't think we should remove it already before starting to process this asset, just in case eg. the queue get's flushed afterwards. 
Also removing the custom setting would require a `save()` -> which I would avoid for performance reasons.

brusch replied:
Sorry, I did got it wrong - fixed, but generic in Asset:update()

kingjia90 replied:
> Sorry, I did got it wrong - fixed, but generic in Asset:update()

The commit about `Asset:update()` wasn't pushed yet in the PR, right? the current on invoke won't allow retry after the file was previously faulty but gets changed with something correct

kingjia90 commented on models/Asset/Video.php:
```diff
@@ -146,13 +146,6 @@ public function getImageThumbnail(array|string|Image\Thumbnail\Config $thumbnail
             return new Video\ImageThumbnail(null); // returns error image
         }
 
-        if (!$this->getCustomSetting('videoWidth') || !$this->getCustomSetting('videoHeight')) {
-            Logger::info('Image thumbnail not yet available, processing is done asynchronously.');
-            $this->addToUpdateTaskQueue();
-
-            return new Video\ImageThumbnail(null); // returns error image
-        }
-
```
mhh, interesting, but i am not fully getting from https://github.com/pimcore/pimcore/pull/11948 the reason behind it, my guess is that we wanted an on-the-fly retry mechanism on potential failure of the Asset Update Task (fatal error, memory leak etc..) ? 

The question would be then, how would it be re-enqueued other than saving?

brusch replied:
That's a valid point - maybe that was the idea behind that. 
But, we'd have the same challenge with the approach from this PR - what happens if one asset fails processing, it get's the custom setting assigned and is stopped from reprocessing forever. 
So I think a command would be nice, to trigger reprocessing for individual asset IDs or paths, because if something is wrong with asset processing, it's most likely a developer that checks and fixes the issue. 

Btw. also saving the asset without changing the data wouldn't trigger reprocessing, see: 
https://github.com/pimcore/pimcore/blob/295879e0fc3b47cf0c406d60e9bf20904bc0e767/models/Asset.php#L609

brusch replied:
Added the command `pimcore:assets:add-to-update-task-queue`

kingjia90 replied:
true, and yes, i meant a dataChanged update, but maybe it would have covered case where the `saveAsset()` is failing (due to db related issues or opensearch? like dead lock or so), so not even ending up storing the custom Setting.
It would be outside from the rollback mechanism, since it's async.

kingjia90 commented on lib/Messenger/Handler/AssetUpdateTasksHandler.php:
```diff
@@ -51,67 +56,77 @@ public function __invoke(AssetUpdateTasksMessage $message): void
 
         $sphericalMetaData = $asset->getSphericalMetaDataFromBackend();
-        if (!empty($sphericalMetaData)) {
+        if ($sphericalMetaData !== $asset->getCustomSetting('SphericalMetaData')) {
```
AFAIK `addToUpdateTaskQueue` is after `getDataChanged` condition, but not sure in case the same file is uplodaded again and has no effectively changes, would still triggers DataChanged (or eg. by copy/paste in the object tree?), the `SsphericalMetaData` might be the same.

Also if former data was valid, but new one is nullish, then should remove the customSetting as it was doing before, something more like

```suggestion
        if (!sphericalMetaData) {
            $asset->removeCustomSetting('SphericalMetaData');
        } elseif ($sphericalMetaData !== $asset->getCustomSetting('SphericalMetaData')) {
            $asset->setCustomSetting('SphericalMetaData', $sphericalMetaData);
        }
```

brusch replied:
Fixed.

kingjia90 commented:
Thank you for the changes!

Had some other follow up comments up there based on the latest changes

On
> * [removed page count processing](https://github.com/pimcore/pimcore/pull/18679/commits/9b96ab1e6675cf10fd2e6368145ddd56124ec02b) in simple backend search -> same here, I don't see any benefit for that, you can't search for that value anyway

indeed, it makes no sense to be called in there as it has no ties/usages with the simple search, must be some legacy leftover (like a sort of riding the wave of the search indexing), https://github.com/pimcore/pimcore/issues/6289#issuecomment-634090270 when count processing was via TmpStore asset-document-conversion, the change LGTM, makes even more sense since it's in an own bundle

brusch committed:
Added command pimcore:assets:add-to-update-task-queue + some further fixes

brusch committed:
Make SonarQube happy

brusch committed:
Apply php-cs-fixer changes

kingjia90 commented on bundles/CoreBundle/src/Command/Asset/AddToUpdateTaskQueueCommand.php:
```diff
@@ -0,0 +1,106 @@
+                null,
+                InputOption::VALUE_OPTIONAL,
+                'only add assets matching the given regex pattern (path + filename), example:  ^/Sample.*urban.jpg$'
+            );
```
thinking would be good to have an option to force retry only on element with `CUSTOM_SETTING_PROCESSING_FAILED:true` (by adding a condition to filter the customSetting column) maybe in some project/instance it would be an use case to have a specific cronjob for it.
Maybe the processing Adapter was temporary broken or fixes were applied or so.

Overall, the PR is turning to look more like an improvement with some minor non-bc-breaking behaviour changes, rather than bugfix, should we change base branch?  As we need to adapt Docs to introduce this new command.

brusch replied:
I'm wondering whether we can filter for that in a SQL query, because actually the custom settings are PHP serialized format. But I agree, if it is easily possible it would be a nice addition. 

Regarding bugfix vs. improvement: I'd still see it as a bug fix because this is the main intention of this PR to fix the broken behavior.

kingjia90 replied:
It's a json array, seems that a LIKE %% https://github.com/pimcore/pimcore/pull/18679/commits/165a1d40f5ba0bc3e602bda3cfb1b65e54cbcfcf#diff-04edf930a619e6e30c664037d7e6169594c2f4e09ad4366d3fe22ceff49ab4d6R101 would work just fine, we only set to true, and key would never exists as false (which is treated as not set)

kingjia90 commented on bundles/CoreBundle/src/Command/Asset/AddToUpdateTaskQueueCommand.php:
```diff
@@ -0,0 +1,106 @@
+                if (in_array($asset->getType(), ['image', 'video', 'document'])) {
+                    $asset->addToUpdateTaskQueue();
+                }
+            }
```
Suggesting to add the types as Listing Condition, to get rid of in_array in the foreach loop, so all limit,offset and load are only on applicable elements, it would also help in saving some memory

```suggestion

/**
 * @internal
 */
#[AsCommand(
    name: 'pimcore:assets:add-to-update-task-queue',
    description: 'Adds assets to update task queue for re-processing (previews, meta-data, ...)',
)]
class AddToUpdateTaskQueueCommand extends AbstractCommand
{
    protected array $types = ['image', 'video', 'document'];
    protected function configure(): void
    {
        $this
            ->addOption(
                'id',
                null,
                InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY,
                'only add assets with this IDs (comma-separated) to the update task queue'
            )
            ->addOption(
                'parent',
                'p',
                InputOption::VALUE_OPTIONAL,
                'only add assets in this folder (ID) to the update task queue'
            )
            ->addOption(
                'path-pattern',
                null,
                InputOption::VALUE_OPTIONAL,
                'only add assets matching the given regex pattern (path + filename), example:  ^/Sample.*urban.jpg$'
            );
    }

    protected function execute(InputInterface $input, OutputInterface $output): int
    {
        $conditionVariables = [];
        $conditions = [];

        if ($input->getOption('parent')) {
            $parent = Asset::getById((int) $input->getOption('parent'));
            if ($parent instanceof Asset\Folder) {
                $conditions[] = "path LIKE '" . Helper::escapeLike($parent->getRealFullPath()) . "/%'";
            } else {
                $this->writeError($input->getOption('parent') . ' is not a valid asset folder ID!');

                return 1;
            }
        }

        if ($ids = $input->getOption('id')) {
            $conditions[] = sprintf('id in (%s)', implode(',', $ids));
        }

        if ($regex = $input->getOption('path-pattern')) {
            $conditions[] = 'CONCAT(`path`, filename) REGEXP ?';
            $conditionVariables[] = $regex;
        }

        $list = new Asset\Listing();

        $condition[] = '`type` IN ('" . implode("','", $this->types) . "')");
        
        $list->setCondition(implode(' AND ', $conditions), $conditionVariables);
        $total = $list->getTotalCount();
        $perLoop = 10;

        for ($i = 0; $i < (ceil($total / $perLoop)); $i++) {
            $list->setLimit($perLoop);
            $list->setOffset($i * $perLoop);
            $assets = $list->load();
            foreach ($assets as $asset) {
                 $asset->addToUpdateTaskQueue();
            }
```

kingjia90 commented on lib/Messenger/Handler/AssetUpdateTasksHandler.php:
```diff
@@ -22,14 +22,19 @@
+    public function __construct(
+        protected LoggerInterface $logger,
+        protected LongRunningHelper $longRunningHelper,
+        protected LockFactory $lockFactory
```
need to add `- '@lock.factory'` or so, under
https://github.com/pimcore/pimcore/blob/5c4a7a21d78355d2af7d91930b3457466ef159e8/bundles/CoreBundle/config/message_handler.yaml#L13

otherwise ends up 
<img width="1902" height="38" alt="image" src="https://github.com/user-attachments/assets/e86c0c27-46e1-4b51-b6d5-a8cd3b1dd486" />

kingjia90 commented on lib/Messenger/Handler/AssetUpdateTasksHandler.php:
```diff
@@ -51,67 +58,74 @@ public function __invoke(AssetUpdateTasksMessage $message): void
         }
 
         $this->longRunningHelper->deleteTemporaryFiles();
+        $this->lockFactory->createLock($asset->getUpdateQueueLockId())->release();
```
This line kind of raises a yellow flag to me, is it right? 
If i got correctly, it is a kind of to instant lock/release to prevent parallel things and/or duplicates?

brusch replied:
Yeah, exactly, this releases the lock, that is being created in `\Pimcore\Model\Asset::addToUpdateTaskQueue()` to prevent duplicates in the queue.

kingjia90 commented on bundles/CoreBundle/src/Command/Asset/AddToUpdateTaskQueueCommand.php:
```diff
@@ -0,0 +1,106 @@
+            $assets = $list->load();
+            foreach ($assets as $asset) {
+                if (in_array($asset->getType(), ['image', 'video', 'document'])) {
+                    $asset->addToUpdateTaskQueue();
```
Should this command be synchronous (or at least as an additional option)?
If yes, instead of `dispatch()` could be `handle()`

brusch replied:
An optional synchronous mode would be nice, yes 👍 
But then we maybe have to think about a different name for the command, because currently it's called "**add to**" 😊

kingjia90 replied:
leaving this out for the moment for an eventual followup, might have timeout issues and if is it too memory hungry for certain adapters etc..

kingjia90 commented:
Other than the inputs above, I've tested it and seems working fine, i feel it is close to be merged.
Sorry for "flooding" this PR with comments 😄

kingjia90 committed:
fix lock service
fix command to filter the type before the listing load
add only retry-failed option to command condition

kingjia90 committed:
Apply php-cs-fixer changes
